Add long press support to LightSwitch and Dimmer devices

esev opened the pull request:
## Description:

This PR exposes the functionality to configure long press actions for the LightSwitch and Dimmer devices.

When combined with #192 & #193 it can be tested with the following script:

```
import logging
import time

import pywemo

WEMO_IP = ''  # Fill this in

def _cb(_device, _type, _params):
    logging.info(
        '%s %s %s', _device.name, _type, _params
    )

logging.basicConfig(level=logging.DEBUG)
sub = pywemo.SubscriptionRegistry()
sub.start()

responder = pywemo.ssdp.DiscoveryResponder(sub.port)
responder.start()

device = pywemo.discovery.device_from_description(
    pywemo.discovery.setup_url_for_address(WEMO_IP, 0), None
)

device.ensure_long_press_virtual_device()
sub.register(device)
sub.on(device, pywemo.subscribe.EVENT_TYPE_LONG_PRESS, _cb)

try:
    time.sleep(999999)
finally:
    responder.stop()
    sub.stop()
```

**Related issue (if applicable):** fixes #61 

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] There is no commented out code in this PR.

esev committed:
Add long press support to LightSwitch and Dimmer devices

Spectre5 commented:
You've been busy!  I'm going to try to check out these 3 PR this evening

Spectre5 commented on tests/ouimeaux_device/test_device.py:
```diff
@@ -369,3 +369,7 @@ def test_setup_successful_2_tries(self, lightspeed, device):
+
+    def test_supports_long_press_is_false(self, lightspeed, device):
+        """Test that the base Device does not have support for long press."""
+        assert device.supports_long_press() == False
```
I think it should be is instead of == here: `assert device.supports_long_press() is False`

esev replied:
Fixed

Spectre5 commented on tests/ouimeaux_device/api/unit/long_press_helpers.py:
```diff
@@ -0,0 +1,70 @@
+import contextlib
+import sqlite3
+import tempfile
+from unittest.mock import create_autospec, patch
```
create_autospec not used

esev replied:
Removed

Spectre5 reviewed (commented):
Looking really cool.  I have not yet tried the 3 PR together to actually test on my devices, but I hope to do that soon. I may wait for the first two to get merged and then test it all together with this PR.

esev committed:
Use an 'is' comparison to test that the value is False

esev committed:
Remove unused create_autospec import

Spectre5 commented:
Just curious, have you tested this with HA as well, or just locally?  I imagine that HA will need a PR after this is fully implemented in pywemo so that the new long press trigger can be used as a trigger for other devices.

esev commented:
> Just curious, have you tested this with HA as well, or just locally? I imagine that HA will need a PR after this is fully implemented in pywemo so that the new long press trigger can be used as a trigger for other devices.

I've been running it in HA for a little over a week.

Future HA integration PR: https://github.com/home-assistant/core/compare/dev...esev:wemo_long_press
Future HA docs: https://github.com/home-assistant/home-assistant.io/compare/next...esev:wemo_long_press

![ink (1)](https://user-images.githubusercontent.com/289218/103606427-376ab600-4ecb-11eb-8532-535452595780.png)

Spectre5 commented:
Very cool.  @pavoni, when do you think you'll have a chance to take a look at these?  I don't want to rush you - just trying to get an estimate.  I'm happy to merge them one-by-one myself, but I figured you'd like to take a look yourself because some parts of these PRs do have some impact on the currently running code (such as the `RequestHandler` doing more duties now).

pavoni commented:
Looks good to me. Feel free to merge.

Will do some overall tests when all the PRs are merged before we release.

Spectre5 commented:
Thanks for all the work on this @esev!

Spectre5 commented:
The script provided in the OP of this PR does work for me for a few lights switches, which is great.  I have not tested in HA though.  @esev, I have never tried making local changes or edits to HA, is there a good source for doing that so I can test your HA PR too?  Probably something I should learn anyway for future testing of other PRs.

esev commented:
My workflow is to use a modified docker container. I pull the official container, add some diff/patch and other files to the container, and then apply the patches to the code in the container.
![Screenshot 2021-01-09 at 11 10 43 AM](https://user-images.githubusercontent.com/289218/104106548-9c2b6500-526b-11eb-8fc2-6020c37b38ab.png)

The files in the `core` folder get copied directly to the respective locations in /usr/src/homeassistant. I use this to add new/custom components to the container. The files in the `patches` folder get further processed with the lines below. This is what I use to modify existing components/code in the docker container.
```
"cd /usr/src/homeassistant && for p in \$(ls /patches/core/*); do /usr/bin/patch --verbose -p 1 < \$p || exit 1; done",
"cd /usr/local/lib/python3.8/site-packages/ && for p in \$(ls /patches/site-packages/*); do /usr/bin/patch -p 1 < \$p || exit 1; done",
```

The diff/patch files are typically made with git, diffing the release branch with my feature branch.

Spectre5 commented:
great, thanks.  I've been too busy with other parts of life to look at this myself so far!

esev commented:
> great, thanks. I've been too busy with other parts of life to look at this myself so far!

All good. Due to the recent [security issue](https://www.home-assistant.io/blog/2021/01/14/security-bulletin/) with HA, I updated to 2021.1.4 today. Rather than patching just the pywemo long-press feature into the HA site-packages, I switched to patching all the pywemo master branch commits since the last release. All the recent changes (dropping py2, updating the generateDS generated code, long-press) are running fine in my HA instance.

```
git checkout master
git diff 0.5.6 pywemo/ > <path>/homeassistant/patches/site-packages/pywemo_master_20210117_a691375.patch
```
